src/base/Quantity.hh

```diff
@@ -208,16 +221,64 @@ swap(Quantity<U, V>& a, Quantity<U, V>& b) noexcept
  * Convert the given quantity into the native Celeritas unit system.
  *
  * \code
- assert(unit_cast(Quantity<real_type, SpeedOfLight>{1})
-        == 2.998e10 * centimeter/second);
+ assert(from_quantity(Quantity<CLight>{1}) == 2.998e10 * centimeter/second);
  * \endcode
  */
 template<class UnitT, class ValueT>
-CELER_CONSTEXPR_FUNCTION auto unit_cast(Quantity<UnitT, ValueT> quant)
-    -> decltype(ValueT() * UnitT::value())
+CELER_CONSTEXPR_FUNCTION auto from_quantity(Quantity<UnitT, ValueT> quant)
+    -> decltype(auto)
+{
+    return quant.value() * UnitT::value();
+}
+
+//! Old spelling of "from_quantity".
+template<class UnitT, class ValueT>
+[[deprecated("use 'from_quantity'")]] CELER_CONSTEXPR_FUNCTION auto
+unit_cast(Quantity<UnitT, ValueT> quant) -> decltype(auto)
 {
     return quant.value() * UnitT::value();
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Create a quantity from a value in the Celeritas unit system.
+ *
+ * This function can be used for defining a constant for use in another unit
+ * system (typically a "natural" unit system for use in physics kernels).
+ *
+ * \code
+ constexpr LightSpeed c = to_quantity<LightSpeed>(constants::c_light);
+ assert(c.value() == 1);
+ * \endcode
+ */
+template<class Q>
+CELER_CONSTEXPR_FUNCTION Q to_quantity(typename Q::value_type value)
+{
+    using value_type = typename Q::value_type;
+    using unit_type  = typename Q::unit_type;
+
+    return Q{value * (value_type{1} / unit_type::value())};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Use the value of a Quantity.
+ *
+ * The redundant unit type in the function signature is to make coupling safer
+ * across different parts of the code and to make the user code more readable.
+ *
+ * \code
+ assert(qvalue<LightSpeed>(LightSpeed{1}) == 1);
+ * \endcode
+ */
+template<class Q, class SrcUnitT, class ValueT>
+CELER_CONSTEXPR_FUNCTION auto qvalue(Quantity<SrcUnitT, ValueT> quant)
+    -> ValueT
+{
+    static_assert(std::is_same<Q, Quantity<SrcUnitT, ValueT>>::value,
+                  "quantity units do not match");
+    return quant.value();
```

Comment:
Since the 'only' costs of using `qvalue` are the extra typing and the `static_assert` (i.e. both 'negligeable'), shouldn't `Quantity::value` be made private (i.e. only allow its use from `qvalue` (which of course needs to be a friend)?

Reply:
Yeah, I considered this, but the extra brackets could make pieces of code that do a lot of math on a lot of different quantities (e.g. the relativistic math in `ParticleTrackView`) a lot uglier than necessary. I think the direct `value` accessor still has a place.

Reply:
humm ... So for example:
```
CELER_FUNCTION units::MevMomentumSq ParticleTrackView::momentum_sq() const
{
    const real_type energy = this->energy().value();
    real_type result = energy * energy + 2 * this->mass().value() * energy;
    CELER_ENSURE(result > 0);
    return units::MevMomentumSq{result};
}
```
On first thought, this looks fine.  It a member function accessing private member, so we already know the type and units are consistent and all is well.  However (and the scenario is more realistic in larger routine and less used code, it is indeed unlikely to happen for *this* case), a future maintainer could be "improving" the code and changing the unit of `this->energy()` and the routine `momentum_sq` would now be silently wrong ....

Reply:
Hopefully we'll catch that in the review process.... but at least this is a step in the right direction :)

Reply:
> Hopefully we'll catch that in the review process....

Technically that is not that straightforward, as this "requires" looking at all usage of `value()` that the author might have forgotten to update (so at the very least the reviewer needs to check the code outside of the PR page).  So where we are using quantities, I think we should (eventually) move toward always using (at least in debug mode) the related checks.